subsys/bluetooth/gatt_dm.c

```diff
@@ -61,69 +70,97 @@ struct bt_gatt_dm {
 /* Currently only one instance is supported */
 static struct bt_gatt_dm bt_gatt_dm_inst;
 
-
-static void *user_data_store(struct bt_gatt_dm *dm,
-			     const void *user_data,
+/* Returns pointer to newly allocated space in a dm->data_chunk */
+static void *user_data_alloc(struct bt_gatt_dm *dm,
 			     size_t len)
 {
 	u8_t *user_data_loc;
-	size_t cur_chunk_id = dm->data_chunk.cur_id;
+	struct data_chunk_item *new_item;
 
-	__ASSERT_NO_MSG(len <= CHUNK_SIZE);
+	/* Round up len to 32 bits to make sure that return pointers are always
+	 * correctly aligned.
+	 */
+	len = (len + DATA_ALIGN - 1) & ~(DATA_ALIGN - 1);
 
-	if (!(dm->data_chunk.chunks[cur_chunk_id])) {
-		dm->data_chunk.chunks[cur_chunk_id] = k_malloc(CHUNK_SIZE);
-		dm->data_chunk.cur_len = 0;
+	__ASSERT_NO_MSG(len <= CHUNK_DATA_SIZE);
 
-		if (!(dm->data_chunk.chunks[cur_chunk_id])) {
-			return NULL;
-		}
-	}
-
-	size_t misalign = dm->data_chunk.cur_len % UUID_ALIGN;
-	if (misalign) {
-		dm->data_chunk.cur_len += UUID_ALIGN - misalign;
-	}
+	if (!dm->data_chunk.cur
+			|| dm->data_chunk.cur_len + len > CHUNK_DATA_SIZE) {
 
-	if (dm->data_chunk.cur_len + len > CHUNK_SIZE) {
-		cur_chunk_id++;
-		dm->data_chunk.cur_id = cur_chunk_id;
-		if ((dm->data_chunk.chunks[cur_chunk_id]) ||
-		    (cur_chunk_id >= ARRAY_SIZE(dm->data_chunk.chunks))) {
-			LOG_ERR("Not enough data chunks");
+		new_item = k_malloc(sizeof(struct data_chunk_item));
+		if (!new_item) {
 			return NULL;
 		}
 
-		return user_data_store(dm, user_data, len);
+		if (dm->data_chunk.cur) {
```

Comment:
I think it would be better to use `slist` instead here.

Reply:
Looks like `slist` will fit nicely here. I will change that.